store/store.go

```diff
@@ -321,34 +321,72 @@ func (s *SnapUbuntuStoreRepository) getPurchasesFromURL(url *url.URL, channel st
 	return purchases, nil
 }
 
-// getPurchases returns user purchases for a specific snap, including in-app.
-func (s *SnapUbuntuStoreRepository) getPurchases(snapID, channel string, auther Authenticator) ([]*purchase, error) {
-	purchasesURL, err := s.purchasesURI.Parse(snapID + "/")
-	if err != nil {
-		return nil, err
+func setMustBuy(snaps []*snap.Info) {
+	for _, info := range snaps {
+		if len(info.Prices) != 0 {
+			info.MustBuy = true
+		}
 	}
+}
 
-	q := purchasesURL.Query()
-	q.Set("include_item_purchases", "true")
-	purchasesURL.RawQuery = q.Encode()
-
-	return s.getPurchasesFromURL(purchasesURL, channel, auther)
+func hasPriced(snaps []*snap.Info) bool {
+	// Search through the list of snaps to see if any are priced
+	for _, info := range snaps {
+		if len(info.Prices) != 0 {
+			return true
+		}
+	}
+	return false
 }
 
-// getAllPurchases returns all user purchases as a map indexed by snap id.
-func (s *SnapUbuntuStoreRepository) getAllPurchases(channel string, auther Authenticator) (map[string][]*purchase, error) {
-	purchases, err := s.getPurchasesFromURL(s.purchasesURI, channel, auther)
+// decorateAllPurchases returns all user purchases as a map indexed by snap id.
+func (s *SnapUbuntuStoreRepository) decoratePurchases(snaps []*snap.Info, channel string, auther Authenticator) error {
+	if auther == nil {
+		// If we can't authenticate, then the non-free snaps must be marked as must buy.
+		setMustBuy(snaps)
+		return nil
+	}
+
+	if !hasPriced(snaps) {
+		// Only look up purchases if there are priced snaps.
+		return nil
+	}
+
+	var err error
+	var purchasesURL *url.URL
+
+	if len(snaps) == 1 {
+		// If we only have a single snap, we should only find the purchases for that snap
+		purchasesURL, err = s.purchasesURI.Parse(snaps[0].SnapID + "/")
+		if err != nil {
+			return err
+		}
+		q := purchasesURL.Query()
+		q.Set("include_item_purchases", "true")
+		purchasesURL.RawQuery = q.Encode()
+	} else {
+		purchasesURL = s.purchasesURI
+	}
+
+	purchases, err := s.getPurchasesFromURL(purchasesURL, channel, auther)
 	if err != nil {
-		return nil, err
+		// If we couldn't get purchase information, then the non-free snaps must be set as must buy.
+		setMustBuy(snaps)
```

Comment:
Code and comment can also go. We're now safe due to the first line and comment in the function.